packages/ui/src/interactions/createFocusVisible.ts

```diff
@@ -0,0 +1,262 @@
+ */
+export function setupGlobalFocusEvents() {
+  if (typeof window === "undefined" || hasSetupGlobalListeners) {
+    return;
+  }
+
+  // Programmatic focus() calls shouldn't affect the current input modality.
+  // However, we need to detect other cases when a focus event occurs without
+  // a preceding user event (e.g. screen reader focus). Overriding the focus
+  // method on HTMLElement.prototype is a bit hacky, but works.
+  const focus = HTMLElement.prototype.focus;
+  HTMLElement.prototype.focus = function (...args) {
+    hasEventBeforeFocus = true;
+    focus.apply(this, args);
+  };
+
+  document.addEventListener("keydown", handleKeyboardEvent, true);
+  document.addEventListener("keyup", handleKeyboardEvent, true);
+  document.addEventListener("click", handleClickEvent, true);
+
+  // Register focus events on the window so they are sure to happen
+  // before React's event listeners (registered on the document).
+  window.addEventListener("focus", handleFocusEvent, true);
+  window.addEventListener("blur", handleWindowBlur, false);
+
+  if (typeof PointerEvent !== "undefined") {
+    document.addEventListener("pointerdown", handlePointerEvent, true);
+    document.addEventListener("pointermove", handlePointerEvent, true);
+    document.addEventListener("pointerup", handlePointerEvent, true);
+  } else {
+    document.addEventListener("mousedown", handlePointerEvent, true);
+    document.addEventListener("mousemove", handlePointerEvent, true);
+    document.addEventListener("mouseup", handlePointerEvent, true);
+  }
+
+  hasSetupGlobalListeners = true;
+}
+
+if (typeof document !== "undefined") {
+  if (document.readyState !== "loading") {
+    setupGlobalFocusEvents();
+  } else {
+    document.addEventListener("DOMContentLoaded", setupGlobalFocusEvents);
+  }
+}
+
+/**
+ * If true, keyboard focus is visible.
+ */
+// export function isFocusVisible(): boolean {
+export function isKeyboardFocusVisible(): boolean {
+  return currentModality !== "pointer";
+}
+
+export function getInteractionModality(): Modality {
+  return currentModality;
+}
+
+export function setInteractionModality(modality: Modality) {
+  currentModality = modality;
+  triggerChangeHandlers(modality);
+}
+
+/**
+ * Keeps state of the current modality.
+ */
+export function useInteractionModality(): Accessor<Modality> {
+  setupGlobalFocusEvents();
+
+  const [modality, setModality] = createSignal(currentModality);
+
+  onMount(() => {
+    const handler = () => {
+      setModality(currentModality);
+    };
+
+    changeHandlers.add(handler);
+
+    onCleanup(() => {
+      changeHandlers.delete(handler);
+    });
+  });
+
+  return modality;
+}
+
+/**
+ * If this is attached to text input component, return if the event is a focus event (Tab/Escape keys pressed) so that
+ * focus visible style can be properly set.
+ */
+function isKeyboardFocusEvent(isTextInput: boolean, modality?: Modality, e?: HandlerEvent) {
+  return !(
+    isTextInput &&
```

Comment:
Can't you check the target to find out if it's a text input, i.e. an instance of HTMLInputElement with a textual type or HTMLTextArea or something inside an element that is contenteditable.

Reply:
Honestly I don't know. I've ported the react code as is because I don't know if major changes will break other primitives I port in the future. With your solution I'm not sure if i should check the target or currentTarget.

Reply:
In input-mask, I've checked either.

Reply:
In react-spectrum the `isTextInput` is used in the `useFocusRing` hook to apply focus ring styles to the wrapping element of a "text field" or the "text field" itself.